build(broccoli): store DiffResult for re-use only if DiffResult

caitp opened the pull request:
One of the non-angular broccoli plugins returns a weird object. We can't
assume that all trees meet the contract that we expect them to meet, so
we do a typecheck before storing the result of the rebuild.

Closes #2662

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/angular/angular/2699)

<!-- Reviewable:end -->

mhevery commented:
Reviewed 1 of 1 files at r3.
Review status: 1 of 35 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/angular/angular/2699)
<!-- Sent from Reviewable.io -->

mhevery commented:
Please rebase to latest master

caitp committed:
Revert "Revert "build(broccoli): allow rebuild() to return DiffResult""

This reverts commit 2c3c23596906acb66b012125ee236c90340c42a7.

caitp commented:
done, `test.unit.dart` + watch() working

caitp committed:
build(broccoli): store DiffResult for re-use only if DiffResult

One of the non-angular broccoli plugins returns a weird object. We can't
assume that all trees meet the contract that we expect them to meet, so
we do a typecheck before storing the result of the rebuild.

Closes #2662

caitp committed:
build(broccoli): don't set stored diffResult to null on use

Doing this would break trees which are used multiple times.

tbosch commented:
Landed in master as 73b3ed18c17bb1e4bd8837e6253d2e16de8208e6
